File: apps/demo-be-rsax/tests/integration/steps/attachment_steps.rs
use axum::body::Body;
use cucumber::{given, then, when};
use http::Request;

use crate::world::{get_req, json_req, AppWorld};

fn multipart_request(
    uri: &str,
    auth: Option<&str>,
    filename: &str,
    content_type: &str,
    data: Vec<u8>,
) -> Request<Body> {
    let boundary = "----TestBoundary123";
    let mut body = format!(
        "--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"{filename}\"\r\nContent-Type: {content_type}\r\n\r\n"
    )
    .into_bytes();
    body.extend_from_slice(&data);
    body.extend_from_slice(format!("\r\n--{boundary}--\r\n").as_bytes());

    let mut builder = Request::builder().method("POST").uri(uri).header(
        "Content-Type",
        format!("multipart/form-data; boundary={boundary}"),
    );
    if let Some(token) = auth {
        builder = builder.header("Authorization", token);
    }
    builder.body(Body::from(body)).unwrap()
}

async fn upload_file(
    world: &mut AppWorld,
    expense_id: &str,
    filename: &str,
    content_type: &str,
    bearer: &str,
    data: Vec<u8>,
) {
    let uri = format!("/api/v1/expenses/{expense_id}/attachments");
    let req = multipart_request(&uri, Some(bearer), filename, content_type, data);
    world.send(req).await.unwrap();
    if world.last_status == 201 {
        world.last_attachment_id = world
            .last_body
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| uuid::Uuid::parse_str(s).ok());
    }
}

#[when(
    regex = r#"alice uploads file "([^"]+)" with content type "([^"]+)" to POST /api/v1/expenses/\{expenseId\}/attachments"#
)]
async fn alice_uploads_file(world: &mut AppWorld, filename: String, content_type: String) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let data = b"fake file content for testing".to_vec();
    upload_file(world, &expense_id, &filename, &content_type, &bearer, data).await;
}

#[given(regex = r#"alice has uploaded file "([^"]+)" with content type "([^"]+)" to the entry"#)]
async fn alice_uploaded_file(world: &mut AppWorld, filename: String, content_type: String) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let data = b"fake file content".to_vec();
    upload_file(world, &expense_id, &filename, &content_type, &bearer, data).await;
}

#[when("alice sends GET /api/v1/expenses/{expenseId}/attachments")]
async fn alice_list_attachments(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let req = get_req(
        &format!("/api/v1/expenses/{expense_id}/attachments"),
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}

#[when("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{attachmentId}")]
async fn alice_delete_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let att_id = world
        .last_attachment_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let req = Request::builder()
        .method("DELETE")
        .uri(&format!(
            "/api/v1/expenses/{expense_id}/attachments/{att_id}"
        ))
        .header("Authorization", &bearer)
        .body(Body::empty())
        .unwrap();
    world.send(req).await.unwrap();
}

#[when("alice uploads an oversized file to POST /api/v1/expenses/{expenseId}/attachments")]
async fn alice_uploads_oversized(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let data = vec![0u8; 11 * 1024 * 1024];
    upload_file(world, &expense_id, "big.jpg", "image/jpeg", &bearer, data).await;
}

#[then(expr = "the response body should contain {int} items in the {string} array")]
async fn response_array_count(world: &mut AppWorld, count: usize, field: String) {
    let actual = world
        .last_body
        .get(&field)
        .and_then(|v| v.as_array())
        .map(|a| a.len())
        .unwrap_or(0);
    assert_eq!(
        actual, count,
        "Expected {count} items in '{field}', got {actual}, body: {}",
        world.last_body
    );
}

#[then(
    regex = r#"the response body should contain an attachment with "([^"]+)" equal to "([^"]+)""#
)]
async fn attachment_with_field(world: &mut AppWorld, field: String, value: String) {
    let attachments = world
        .last_body
        .get("attachments")
        .and_then(|v| v.as_array());
    let found = attachments
        .map(|arr| {
            arr.iter().any(|att| {
                att.get(&field)
                    .and_then(|v| v.as_str())
                    .map(|s| s == value.as_str())
                    .unwrap_or(false)
            })
        })
        .unwrap_or(false);
    assert!(
        found,
        "Expected attachment with '{field}' = '{value}' in: {}",
        world.last_body
    );
}

#[then("the response body should contain an error message about file size")]
async fn error_file_size(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected file size error, got: {}",
        world.last_body
    );
}

// Bob steps for cross-user ownership tests

#[given(
    regex = r#"bob has created an entry with body \{ "amount": "25\.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense" \}"#
)]
async fn bob_created_entry(world: &mut AppWorld) {
    let login_req = json_req(
        "POST",
        "/api/v1/auth/login",
        r#"{"username": "bob", "password": "Str0ng#Pass2"}"#,
        None,
    );
    world.send(login_req).await.unwrap();
    if world.last_status == 200 {
        world.bob_auth_token = world
            .last_body
            .get("access_token")
            .and_then(|v| v.as_str())
            .map(String::from);
        let bob_bearer = world.bob_bearer();
        let req = json_req(
            "POST",
            "/api/v1/expenses",
            r#"{"amount": "25.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense"}"#,
            Some(&bob_bearer),
        );
        world.send(req).await.unwrap();
        if world.last_status == 201 {
            world.bob_expense_id = world
                .last_body
                .get("id")
                .and_then(|v| v.as_str())
                .and_then(|s| uuid::Uuid::parse_str(s).ok());
        }
    }
}

#[when(
    regex = r#"alice uploads file "([^"]+)" with content type "([^"]+)" to POST /api/v1/expenses/\{bobExpenseId\}/attachments"#
)]
async fn alice_uploads_to_bob_expense(
    world: &mut AppWorld,
    filename: String,
    content_type: String,
) {
    let bearer = world.bearer();
    let expense_id = world
        .bob_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let data = b"file content".to_vec();
    upload_file(world, &expense_id, &filename, &content_type, &bearer, data).await;
}

#[when("alice sends GET /api/v1/expenses/{bobExpenseId}/attachments")]
async fn alice_list_bob_attachments(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .bob_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let req = get_req(
        &format!("/api/v1/expenses/{expense_id}/attachments"),
        Some(&bearer),
    );
    world.send(req).await.unwrap();
}

#[when("alice sends DELETE /api/v1/expenses/{bobExpenseId}/attachments/{attachmentId}")]
async fn alice_delete_bob_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .bob_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let att_id = world
        .last_attachment_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let req = Request::builder()
        .method("DELETE")
        .uri(&format!(
            "/api/v1/expenses/{expense_id}/attachments/{att_id}"
        ))
        .header("Authorization", &bearer)
        .body(Body::empty())
        .unwrap();
    world.send(req).await.unwrap();
}

#[when("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{randomAttachmentId}")]
async fn alice_delete_nonexistent_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = world
        .last_expense_id
        .map(|id| id.to_string())
        .unwrap_or_default();
    let random_id = uuid::Uuid::new_v4();
    let req = Request::builder()
        .method("DELETE")
        .uri(&format!(
            "/api/v1/expenses/{expense_id}/attachments/{random_id}"
        ))
        .header("Authorization", &bearer)
        .body(Body::empty())
        .unwrap();
    world.send(req).await.unwrap();
}

File: apps/demo-be-python-fastapi/tests/integration/steps/attachment_steps.py
"""BDD step definitions for attachment feature."""

import json
import uuid

from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT
from tests.integration.service_client import FakeResponse, ServiceClient

scenarios(str(GHERKIN_ROOT / "expenses" / "attachments.feature"))

_PASSWORD = "Str0ng#Pass1"
_PASSWORD2 = "Str0ng#Pass2"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_attach(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


@given(
    parsers.parse("alice has created an entry with body {body}"),
    target_fixture="created_expense",
)
def alice_create_attachment_entry(client: ServiceClient, alice_tokens: dict, body: str) -> dict:
    data = json.loads(body)
    resp = client.post_expense(f"Bearer {alice_tokens['access_token']}", data)
    assert resp.status_code == 201, f"Create entry failed: {resp.text}"
    return resp.json()


@given(
    parsers.parse("bob has created an entry with body {body}"),
    target_fixture="bob_expense",
)
def bob_create_entry(client: ServiceClient, registered_user: dict, body: str) -> dict:
    # registered_user is bob (last registration in the scenario)
    bob_tokens = client.login_user("bob", _PASSWORD2)
    data = json.loads(body)
    resp = client.post_expense(f"Bearer {bob_tokens['access_token']}", data)
    assert resp.status_code == 201
    return resp.json()


@given(
    parsers.parse(
        'alice has uploaded file "{filename}" with content type "{content_type}" to the entry'
    ),
    target_fixture="uploaded_attachment",
)
def alice_upload_attachment_given(
    client: ServiceClient,
    alice_tokens: dict,
    created_expense: dict,
    filename: str,
    content_type: str,
) -> dict:
    file_content = b"dummy file content"
    resp = client.post_attachment(
        created_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
        filename,
        content_type,
        file_content,
    )
    assert resp.status_code == 201, f"Upload failed: {resp.text}"
    return resp.json()


# --- When steps ---


@when(
    parsers.parse(
        'alice uploads file "{filename}" with content type "{content_type}" to POST /api/v1/expenses/{{expenseId}}/attachments'  # noqa: E501
    ),
    target_fixture="response",
)
def alice_upload_file(
    client: ServiceClient,
    alice_tokens: dict,
    created_expense: dict,
    filename: str,
    content_type: str,
) -> FakeResponse:
    file_content = b"dummy file content"
    return client.post_attachment(
        created_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
        filename,
        content_type,
        file_content,
    )


@when(
    parsers.parse(
        'alice uploads file "{filename}" with content type "{content_type}" to POST /api/v1/expenses/{{bobExpenseId}}/attachments'  # noqa: E501
    ),
    target_fixture="response",
)
def alice_upload_to_bob_expense(
    client: ServiceClient,
    alice_tokens: dict,
    bob_expense: dict,
    filename: str,
    content_type: str,
) -> FakeResponse:
    file_content = b"dummy file content"
    return client.post_attachment(
        bob_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
        filename,
        content_type,
        file_content,
    )


@when(
    "alice uploads an oversized file to POST /api/v1/expenses/{expenseId}/attachments",
    target_fixture="response",
)
def alice_upload_oversized(
    client: ServiceClient, alice_tokens: dict, created_expense: dict
) -> FakeResponse:
    big_content = b"x" * (11 * 1024 * 1024)
    return client.post_attachment(
        created_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
        "large.pdf",
        "application/pdf",
        big_content,
    )


@when("alice sends GET /api/v1/expenses/{expenseId}/attachments", target_fixture="response")
def alice_list_attachments(
    client: ServiceClient, alice_tokens: dict, created_expense: dict
) -> FakeResponse:
    return client.get_attachments(
        created_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
    )


@when("alice sends GET /api/v1/expenses/{bobExpenseId}/attachments", target_fixture="response")
def alice_list_bob_attachments(
    client: ServiceClient, alice_tokens: dict, bob_expense: dict
) -> FakeResponse:
    return client.get_attachments(
        bob_expense["id"],
        f"Bearer {alice_tokens['access_token']}",
    )


@when(
    "alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{attachmentId}",
    target_fixture="response",
)
def alice_delete_attachment(
    client: ServiceClient,
    alice_tokens: dict,
    created_expense: dict,
    uploaded_attachment: dict,
) -> FakeResponse:
    return client.delete_attachment(
        created_expense["id"],
        uploaded_attachment["id"],
        f"Bearer {alice_tokens['access_token']}",
    )


@when(
    "alice sends DELETE /api/v1/expenses/{bobExpenseId}/attachments/{attachmentId}",
    target_fixture="response",
)
def alice_delete_bob_attachment(
    client: ServiceClient,
    alice_tokens: dict,
    bob_expense: dict,
    uploaded_attachment: dict,
) -> FakeResponse:
    return client.delete_attachment(
        bob_expense["id"],
        uploaded_attachment["id"],
        f"Bearer {alice_tokens['access_token']}",
    )


@when(
    "alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{randomAttachmentId}",
    target_fixture="response",
)
def alice_delete_nonexistent_attachment(
    client: ServiceClient, alice_tokens: dict, created_expense: dict
) -> FakeResponse:
    random_id = str(uuid.uuid4())
    return client.delete_attachment(
        created_expense["id"],
        random_id,
        f"Bearer {alice_tokens['access_token']}",
    )


# --- Then steps ---


@then(parsers.parse('the response body should contain 2 items in the "attachments" array'))
def check_two_attachments(response: FakeResponse) -> None:
    body = response.json()
    attachments = body.get("attachments", [])
    assert len(attachments) == 2, f"Expected 2 attachments, got {len(attachments)}: {body}"


@then(
    parsers.parse(
        'the response body should contain an attachment with "filename" equal to "{filename}"'
    )
)
def check_attachment_filename(response: FakeResponse, filename: str) -> None:
    body = response.json()
    attachments = body.get("attachments", [])
    assert any(a.get("filename") == filename for a in attachments), (
        f"Attachment '{filename}' not found in: {attachments}"
    )


@then(parsers.parse('the response body should contain "content_type" equal to "{content_type}"'))
def check_attachment_content_type(response: FakeResponse, content_type: str) -> None:
    body = response.json()
    assert "content_type" in body, f"'content_type' not in response: {body}"
    assert body["content_type"] == content_type, (
        f"Expected content_type={content_type!r}, got {body['content_type']!r}"
    )


@then(parsers.parse('the response body should contain "filename" equal to "{filename}"'))
def check_filename(response: FakeResponse, filename: str) -> None:
    body = response.json()
    assert "filename" in body, f"'filename' not in response: {body}"
    assert body["filename"] == filename, (
        f"Expected filename={filename!r}, got {body['filename']!r}"
    )


@then('the response body should contain a validation error for "file"')
def check_file_validation_error(response: FakeResponse) -> None:
    body = response.json()
    body_str = json.dumps(body).lower()
    assert any(kw in body_str for kw in ["file", "media", "type", "unsupported"]), (
        f"Expected file validation error, got: {body}"
    )
